File: FloRelPy/node.py
from hash import Hash
from utils import Status
from init import initialize
from multiprocessing import Lock, Queue

class Node:
	def __init__(self, id,
		neighbor_count=0,
		device_count=0
	):
		self.id = id

		self.neighbor_count = neighbor_count
		self.device_count = device_count

		self.global_results = Queue()

		self.device_hash = Hash(self.device_count)
		self.results_hash = Hash(self.neighbor_count)
		self.receive_hash = Hash(self.neighbor_count)
		self.neighbor_map = Hash(self.neighbor_count)

		self.mutex = Lock()
		self.status = Status(running=True)

	def get_status(self):
		with self.mutex:
			return self.status

	def fetch_top_result(self):
		return self.global_results.get()

	def run(self):
		while self.status.running:
			pass
		return True

	def go_online(self):
		if initialize(self):
			return self.run()
		else:
			print("Node failed to initialize")
		return False

File: FloRelC-sx1278/src/node.c
#include "../includes/florel.h"
#include "../includes/utils.h"
#include "../includes/node.h"

void free_node(t_node *node) {
	if (node) {
		free_hash(node->device_hash);
		free_hash(node->results_hash);
		free_hash(node->receive_hash);
		free_hash(node->neighbor_map);
		
		free_queue(node->global_results);
		free(node);
	}
}

t_status *get_status(t_node *node) {
	t_status *result;

	result = NULL;
	if (!pthread_mutex_trylock(&node->lock.lock)) {
		result = &node->status;
		pthread_mutex_unlock(&node->lock.lock);
	}

	return (result);
}

t_node *new_node(char **argv) {
	// TODO needs devices
	t_node *node;
	unsigned int id;

	if (!(node = (t_node*)calloc(1, sizeof(t_node))))
		return (NULL);
	id = atoi(argv[1]);
	node->id = id;

	node->status = *new_status(true);

	return (node);
}